### PR TITLE
1script

### PR DESCRIPTION
**Pull request by WackerO**
Many thanks to contributing to qbic-pipelines!

This PR combines the pipeline code from the different R scripts and modules into a single script/module; all computation is now done in the report.rmd. Also, the multiQC is now an optional input.

## PR checklist
 - [x] This comment contains a description of changes (with reason)
 - [x] If you've fixed a bug or added code that should be tested, add tests!
 - [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
 - [x] Documentation in `docs` is updated
 - [x] `CHANGELOG.md` is updated
 - [x] `README.md` is updated

**Learn more about contributing:** https://github.com/qbic-pipelines/rnadeseq/tree/master/.github/CONTRIBUTING.md

### CODE REVIEW

**Comment by WackerO** [comment]
TODO: Check if all recent changes are implemented in 1script as well

File: assets/RNAseq_report.Rmd [hunk 1]
```diff
@@ -12,27 +12,163 @@ output:
...
+    revision: ''
+    pathway_analysis: ''
+    rlog: ''
+    multiqc: ''
 
 # Author: Silvia Morini, Gisela Gabernet, Simon Heumos
 ---
 
-```{r, bootstrap.show.message=FALSE, echo=FALSE, message=FALSE}
+```{r, echo=FALSE, message=FALSE, warning=FALSE, results = 'hide'}
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:67]:
```suggestion
```{r param_validation, echo=FALSE, message=FALSE, warning=FALSE, results = 'hide'}
```

File: assets/RNAseq_report.Rmd [hunk 2]
```diff
@@ -12,27 +12,163 @@ output:
...
+"extrafont",
+"limma",
+"tximeta",
+"tximport",
+"SummarizedExperiment",
+"impute"
+), library, character.only=T))
+```
+
+```{r, echo=FALSE, message=FALSE, warning=FALSE, results = 'hide'}
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:157]:
```suggestion
```{r create_outdirs, echo=FALSE, message=FALSE, warning=FALSE, results = 'hide'}
```

File: assets/RNAseq_report.Rmd [hunk 3]
```diff
@@ -134,10 +256,66 @@ pval_threshold <- 0.05  #set here the non-logarithmized threshold for the p valu
 pval_text <- as.character(pval_threshold)
 ```
 
-```{r setup, echo=FALSE, message=FALSE}
+```{r setup, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:259]:
```suggestion
```{r read_metadata, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
```

File: assets/RNAseq_report.Rmd [hunk 4]
```diff
@@ -341,19 +834,55 @@ The original plot can be downloaded [here](./differential_gene_expression/plots/
...
+    ylab(paste0("PC2: ",percentVar[2], "% variance")) +
+    theme(legend.title = element_blank()) +
+    coord_fixed() +
+    theme_bw() +
+    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
+suppressMessages(ggsave(plot = pca, filename = "differential_gene_expression/plots/PCA_plot.pdf", device = "pdf", dpi = 300))
+suppressMessages(ggsave(plot = pca, filename = "differential_gene_expression/plots/PCA_plot.svg", device = "svg", dpi = 150))
+suppressMessages(ggsave(plot = pca, filename = "differential_gene_expression/plots/PCA_plot.png", device = "png", dpi = 150))
+########################### PCA PLOT with batch-corrected data ############
+if(!isNot(params$batch_effect)){
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:853]: Can't this be done with an `eval` setting to at the beginning of the block?

File: assets/RNAseq_report.Rmd [hunk 5]
```diff
@@ -341,19 +834,55 @@ The original plot can be downloaded [here](./differential_gene_expression/plots/
 
 \
 
-```{r, echo=F, out.width="80%", dpi=1200, fig.align='center'}
-knitr::include_graphics(paste0(wd,"/differential_gene_expression/plots/PCA_plot.svg"))
+```{r PCA_plots, echo=F, warning=F}
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:837]:
```suggestion
```{r PCA_plots, echo=F, eval=!params$batch_effect, warning=F}
```

File: assets/RNAseq_report.Rmd (hunk 4 above)
Comment (ggabernet) [reply assets/RNAseq_report.Rmd:853]: Then this if clause is not necessary, and this code can be moved to the next block (PCA_ifBatch_plot)

File: assets/RNAseq_report.Rmd [hunk 6]
```diff
@@ -364,26 +893,222 @@ This visualizes how similar the samples are to each other according to gene expr
...
+    contrast_names <- append(contrast_names, colnames(contrasts))
+}
+if (!(isNot(params$path_contrast_list))) {
+    contrasts <- read.table(params$path_contrast_list, sep="\t", header=T, colClasses = "character")
+    write.table(contrasts, file="differential_gene_expression/metadata/contrast_list.tsv", quote=F)
+    ## Contrast calculation for contrast list
+    for (i in c(1:nrow(contrasts))) {
+        cont <- as.character(contrasts[i,])
+        contname <- paste0(cont[1], "_", cont[2], "_vs_", cont[3])
+    # TODO: add checks if provided contrast_names and factors are in metadata
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:973]: Is this done? Otherwise, could you open an issue about it?

File: assets/RNAseq_report.Rmd [hunk 7]
```diff
@@ -364,26 +893,222 @@ This visualizes how similar the samples are to each other according to gene expr
...
+#write to file
+write.table(DE_genes_final_table, "differential_gene_expression/final_gene_table/final_DE_gene_list.tsv", append = FALSE, quote = FALSE, sep = "\t",eol = "\n", na = "NA", dec = ".", row.names = F,  col.names = T, qmethod = c("escape", "double"))
 ```
+<!-- contrast calculation end -->
 
-```{r contrasts_block, echo=FALSE, results='asis'}
-for (i in 1:length(contrnames)){
-    cat(paste("* ", contrnames[i], "\n"))
+```{r print_contrasts, echo=FALSE, results='asis'}
+#TODO:
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:1102]: ?

File: assets/RNAseq_report.Rmd [hunk 8]
```diff
@@ -407,6 +1131,7 @@ The table for download can be found [here](./differential_gene_expression/final_
 Additional tables for all individual contrasts are provided [here](./differential_gene_expression/DE_genes_tables).
 
 <!-- Show table here removed as I dont like showing only top 20 genes.
+TODO Is this still necessary?
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:1134]: If the table here is not shown (fine by me, please clarify with @qbicStefanC , then the text above should be adapted), as no list is shown.

File: assets/RNAseq_report.Rmd [hunk 9]
```diff
@@ -459,11 +1176,12 @@ Here, volcano plots for all considered contrasts are shown. Hover over the dots
...
     pg <- ggplot(DE_all, aes(x=log2FoldChange, y=logpval, text=paste("Gene: ", gene_name, "<br>", "Log2FC: ", formatC(log2FoldChange, digits=2)))) +
-            geom_point(alpha=0.5, aes(color=ifelse(abs(DE_all$log2FoldChange) > params$log_FC &  DE_all$logpval > -log10(pval_threshold), "Differentially expressed genes", "Non-differentially expressed genes"))) +
+            #this goes after alpha=0.5:
+            geom_point(alpha=0.5, aes(color=ifelse(abs(DE_all$log2FoldChange) > params$log_FC_threshold &  DE_all$logpval > -log10(pval_threshold), "Differentially expressed genes", "Non-differentially expressed genes"))) +
             geom_hline(yintercept = 16, linetype= "dashed", size = 0.2, color = "grey") +
             geom_hline(yintercept = -log10(pval_threshold), size = 0.2, color = "grey") +
-            geom_vline(xintercept = log_FC_num, size = 0.2, color = "grey") +
-            geom_vline(xintercept = -log_FC_num, size = 0.2, color = "grey") +
+            geom_vline(xintercept = 0.05, size = 0.2, color = "grey") +
+            geom_vline(xintercept = -0.05, size = 0.2, color = "grey") +
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:1184]: here refer also to the `pvalue` threshold parameter, no hard coding of number.

File: assets/RNAseq_report.Rmd [hunk 10]
```diff
@@ -472,7 +1190,14 @@ Here, volcano plots for all considered contrasts are shown. Hover over the dots
...
     pg[['x']][['layout']][['annotations']][[1]][['y']] <- -0.05
-     # This adds a save-to-svg button to the plot menubar
+
+    # svg path for the little svg format icon displayed in the volcano plot for svg export; based on https://www.svgviewer.dev/:
+    svg_icon_path <- 'M228 319.5H12C5.64873 319.5 0.5 314.351 0.5 308V12C0.5 5.64873 5.64873 0.5 12 0.5H159.175C159.571 0.5 159.952 0.656748 160.233 0.936021L239.057 79.2681C239.341 79.5497 239.5 79.9326 239.5 80.3321V308C239.5 314.351 234.351 319.5 228 319.5Z'
+    svg_icon_fill <- 'darkgrey'
+    svg_icon_transform <- 'scale(0.07)'
+    svg_icon_viewbox <- '0 0 240 320'
+    # This adds a save-to-svg button to the plot menubar
+    #TODO: check if svg works
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:1200]: Is this solved? Otherwise could you open an issue?

File: assets/RNAseq_report.Rmd [hunk 11]
```diff
@@ -514,6 +1290,280 @@ cat("Plots for the [genes of interest](./differential_gene_expression/metadata/g
...
+invisible(lapply(c(
+"gprofiler2",
+"reshape2",
+"pathview",
+"AnnotationDbi",
+"optparse",
+"org.Mm.eg.db",
+"org.Hs.eg.db"
+), library, character.only=T))
+# Need to load library for your species! Or provide as parameter to biocmanager
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:1306]: I thought this is solved by your `install(params$species_libraries)` code below already.

File: assets/RNAseq_report.Rmd [hunk 12]
```diff
@@ -514,6 +1290,280 @@ cat("Plots for the [genes of interest](./differential_gene_expression/metadata/g
...
+
+# For each contrast do pathway analysis
+
+# TODO debugging summary table
+# Defining summary variables
+# contrast <- c()
+# number_DE_genes <- c()
+# number_enriched_pathways <- c()
+# list_DE_genes <- list()
+# list_enriched_pathways <- list()
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:1355]: Can this be removed?

File: assets/RNAseq_report.Rmd [hunk 13]
```diff
@@ -514,6 +1290,280 @@ cat("Plots for the [genes of interest](./differential_gene_expression/metadata/g
...
+                sep="\t", quote = F, col.names = T, row.names = F)
+
+    # TODO debugging summary table
+    # Collecting summary variables
+    # contrast <- append(contrast, fname)
+    # number_DE_genes <- append(number_DE_genes, length(DE_genes$Ensembl_ID))
+    # number_enriched_pathways <- append(number_enriched_pathways, summary(as.factor(pathway_gostres_table$source)))
+    # list_DE_genes <- append(list_DE_genes, DE_genes$Ensembl_ID)
+    # list_enriched_pathways <- append(list_enriched_pathways, pathway_gostres_table$term_id)
+    # Printing summary variables
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:1420]: Maybe this can be removed now too

File: assets/RNAseq_report.Rmd [hunk 14]
```diff
@@ -514,6 +1290,280 @@ cat("Plots for the [genes of interest](./differential_gene_expression/metadata/g
...
+            }
+        }
+    }
+}
+
+# Writing DE gene summary table
+# df_summary <- data.frame(Contrast = contrast, Number_DE_genes = number_DE_genes, Number_enriched_pathways = number_enriched_pathways, DE_genes_list = list_DE_genes, Enriched_pathways_list = list_enriched_pathways)
+# write.table(df_summary,
+#         file = paste0("pathway_analysis", "/", fname, "/", fname, "_pathway_enrichment_summary.tsv"),
+#         sep="\t", quote = F, col.names = T, row.names = F)
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:1519]: Would it be useful to still save this table?

File: assets/RNAseq_report.Rmd [hunk 15]
```diff
@@ -530,9 +1580,7 @@ Hover over the dots to reveal the pathway names.
 The table below provides more detail on all enriched pathways."))
 ```
 
-```{r, eval=params$pathway_analysis, out.height = "100%", out.width = "100%", warning=FALSE, message=FALSE}
-    library(gprofiler2)
-    #library(rlist)
+```{r gost, eval=params$pathway_analysis, out.height = "100%", out.width = "100%", warning=FALSE, message=FALSE}
     #TODO: check if here, only the filtered genes are necessary
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:1584]: Is it solved? Otherwise convert to issue please :)

File: assets/RNAseq_report.Rmd [hunk 16]
```diff
@@ -563,15 +1613,15 @@ The table below provides more detail on all enriched pathways."))
     path_enrich <- as.data.frame(gostres$result)
     if (nrow(path_enrich) > 0){
         pg2 <- gostplot(gostres, capped=T, interactive=T)
-        pg2
+    # TODO   pg2
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:1616]: ?

File: bin/Execute_report.R [hunk 17]
```diff
@@ -4,23 +4,32 @@ library(rmarkdown)
 library(optparse)
 
 option_list = list(
```
Comment (ggabernet) [inline bin/Execute_report.R:6]: It would be really nice to group the options here in the same way and order as they are read in the report file

File: bin/Execute_report.R [hunk 18]
```diff
@@ -36,17 +45,25 @@ if(!is.null(opt$genelist)){
 
 rmarkdown::render(opt$report, output_file = opt$output, knit_root_dir = wd, output_dir = wd,
```
Comment (ggabernet) [inline bin/Execute_report.R:46]: here also the same parameter grouping and order would be really helpful

File: nextflow.config [hunk 19]
```diff
@@ -184,7 +185,7 @@ manifest {
     description     = 'Differential gene expression analysis and pathway analysis of RNAseq data'
     mainScript      = 'main.nf'
     nextflowVersion = '!>=21.10.3'
-    version = '2.0.1'
+    version = '2.0.1' //TODO: does this need to be dev?
```
Comment (ggabernet) [inline nextflow.config:188]:
```suggestion
    version = '2.1.0dev'
```

Comment (ggabernet) [reply nextflow.config:188]: yes, new expected version + dev at the end

File: workflows/rnadeseq.nf [hunk 20]
```diff
@@ -51,8 +51,8 @@ ch_genes = Channel.fromPath(params.genelist)
 //
 // MODULE: Loaded from modules/local/
 //
-include { DESEQ2 } from '../modules/local/deseq2'
-include { PATHWAY_ANALYSIS } from '../modules/local/pathway_analysis'
+//include { DESEQ2 } from '../modules/local/deseq2'
+//include { PATHWAY_ANALYSIS } from '../modules/local/pathway_analysis'
```
Comment (ggabernet) [inline workflows/rnadeseq.nf:55]:
```suggestion
```

File: workflows/rnadeseq.nf [hunk 21]
```diff
@@ -147,9 +120,9 @@ workflow RNADESEQ {
 //Here I have to change the params?
 
 /*
-//TODO: Remove multiqc and also edit /lib/nfcoretemplate to add deseq report instead of multiqc
+//TODO: Edit /lib/nfcoretemplate to add deseq report instead of multiqc
```
Comment (ggabernet) [inline workflows/rnadeseq.nf:123]: Maybe convert to issue so we don't forget

**Review by ggabernet — COMMENTED** [review]
Hi Oskar, great job with the huge code restructuring. I added some comments, especially to try to improve readability and structuring of the code, also a couple more things. Let me know if something is unclear!

File: assets/RNAseq_report.Rmd [hunk 22]
```diff
@@ -152,14 +330,287 @@ if(file.exists(QAfile)){
...
+################## RUN DESEQ2 ######################################
+#Apply relevel if provided to metadata
+if (!isNot(params$path_relevel)) {
+    relevel_table <- read.table(params$path_relevel, sep="\t", header = T, colClasses = "character")
+    write.table(relevel_table, file="differential_gene_expression/metadata/relevel.tsv")
+    for (i in c(1:nrow(relevel_table))) {
+        relev <- relevel_table[i,]
+        cds[[paste(relev[1])]] <- relevel(cds[[paste(relev[1])]], paste(relev[2]))
+    }
+}
```
[the remark this answers is not in the document]
Comment (WackerO) [reply assets/RNAseq_report.Rmd]: I believe it should be fine; I don't see any redefinition of `cds` after the relevel code (apart from the transformations)

File: assets/RNAseq_report.Rmd [hunk 23]
```diff
@@ -152,14 +330,287 @@ if(file.exists(QAfile)){
...
+# Run DESeq function
+if (params$input_type == "featurecounts") {
+    cds <- DESeqDataSetFromMatrix( countData =count.table, colData =metadata, design = eval(parse(text=as.character(design[[1]]))))
+    cds <- DESeq(cds,  parallel = FALSE)
+} else if (params$input_type %in% c("rsem", "salmon")) {
+    ## Create a dataframe which consists of both the gene id and the transcript name
+    gtf <- rtracklayer::import(params$path_gtf)
+    gtf <- as.data.frame(gtf, header=T)
+    # TODO: For some gtf files, transcript_id does not work!!
```
[the remark this answers is not in the document]
Comment (WackerO) [reply assets/RNAseq_report.Rmd:345]: Hm, so this would have to be looked at when the error appears again (i.e. when someone tries a different gtf and it does not work). Or should I create an issue to which people can then respond? I haven't triggered that error since I implemented the gtf stuff...

File: assets/RNAseq_report.Rmd (hunk 12 above)
Comment (WackerO) [reply assets/RNAseq_report.Rmd:1344]: Hm, I don't know :'D This part was already commented out when I started working on deseq

File: assets/RNAseq_report.Rmd [hunk 24]
```diff
@@ -364,26 +897,221 @@ This visualizes how similar the samples are to each other according to gene expr
...
 
 # Differential expression analysis
 
 To perform differential gene expression analysis, the gene expression data between two sample groups is compared.
 These comparisons are called _contrasts_. For this analysis, the following contrasts or comparisons were made:
 
-```{r contrasts, echo=FALSE, message=FALSE, bootstrap.show.message=FALSE, include=FALSE}
-df_contr <- read.table(params$path_contrasts, header=FALSE, sep="\t")
-contrnames = df_contr$V1
+<!-- contrast calculation start -->
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:928]: The idea of the Rmarkdown report, is that each section contains the code that is needed to reproduce the plots. Could you please move the code that calculates the DESEq2 objects here? That would be the section that it belongs to

File: assets/RNAseq_report.Rmd [hunk 25]
```diff
@@ -459,11 +1179,12 @@ Here, volcano plots for all considered contrasts are shown. Hover over the dots
...
     pg <- ggplot(DE_all, aes(x=log2FoldChange, y=logpval, text=paste("Gene: ", gene_name, "<br>", "Log2FC: ", formatC(log2FoldChange, digits=2)))) +
-            geom_point(alpha=0.5, aes(color=ifelse(abs(DE_all$log2FoldChange) > params$log_FC &  DE_all$logpval > -log10(pval_threshold), "Differentially expressed genes", "Non-differentially expressed genes"))) +
+            #this goes after alpha=0.5:
+            geom_point(alpha=0.5, aes(color=ifelse(abs(DE_all$log2FoldChange) > params$log_FC_threshold &  DE_all$logpval > -log10(pval_threshold), "Differentially expressed genes", "Non-differentially expressed genes"))) +
             geom_hline(yintercept = 16, linetype= "dashed", size = 0.2, color = "grey") +
             geom_hline(yintercept = -log10(pval_threshold), size = 0.2, color = "grey") +
-            geom_vline(xintercept = log_FC_num, size = 0.2, color = "grey") +
-            geom_vline(xintercept = -log_FC_num, size = 0.2, color = "grey") +
+            geom_vline(xintercept = 0.05, size = 0.2, color = "grey") +
+            geom_vline(xintercept = -0.05, size = 0.2, color = "grey") +
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:1187]: Why is this here `0.05` instead of the log_FC threshold?

File: assets/RNAseq_report.Rmd [hunk 26]
```diff
@@ -514,6 +1292,281 @@ cat("Plots for the [genes of interest](./differential_gene_expression/metadata/g
...
+
+# For each contrast do pathway analysis
+
+# TODO debugging summary table
+# Defining summary variables
+# contrast <- c()
+# number_DE_genes <- c()
+# number_enriched_pathways <- c()
+# list_DE_genes <- list()
+# list_enriched_pathways <- list()
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:1358]:
```suggestion
```

Comment (ggabernet) [reply assets/RNAseq_report.Rmd:1358]: yes then it can be :)

File: assets/RNAseq_report.Rmd [hunk 27]
```diff
@@ -514,6 +1292,281 @@ cat("Plots for the [genes of interest](./differential_gene_expression/metadata/g
...
+                sep="\t", quote = F, col.names = T, row.names = F)
+
+    # TODO debugging summary table
+    # Collecting summary variables
+    # contrast <- append(contrast, fname)
+    # number_DE_genes <- append(number_DE_genes, length(DE_genes$Ensembl_ID))
+    # number_enriched_pathways <- append(number_enriched_pathways, summary(as.factor(pathway_gostres_table$source)))
+    # list_DE_genes <- append(list_DE_genes, DE_genes$Ensembl_ID)
+    # list_enriched_pathways <- append(list_enriched_pathways, pathway_gostres_table$term_id)
+    # Printing summary variables
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:1423]:
```suggestion
```

File: assets/RNAseq_report.Rmd [hunk 28]
```diff
@@ -514,6 +1292,281 @@ cat("Plots for the [genes of interest](./differential_gene_expression/metadata/g
...
+            }
+        }
+    }
+}
+
+# Writing DE gene summary table
+# df_summary <- data.frame(Contrast = contrast, Number_DE_genes = number_DE_genes, Number_enriched_pathways = number_enriched_pathways, DE_genes_list = list_DE_genes, Enriched_pathways_list = list_enriched_pathways)
+# write.table(df_summary,
+#         file = paste0("pathway_analysis", "/", fname, "/", fname, "_pathway_enrichment_summary.tsv"),
+#         sep="\t", quote = F, col.names = T, row.names = F)
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:1522]:
```suggestion
```

**Review by ggabernet — COMMENTED** [review]
Hi @WackerO, I have some more comments, the rest looks good :)

File: assets/RNAseq_report.Rmd [hunk 29]
```diff
@@ -123,21 +248,83 @@ Auf der Morgenstelle 10, 72076 Tuebingen
...
+system(paste("mv ",params$path_metadata," differential_gene_expression/metadata/metadata.tsv",sep=""))
+qbicCodes <- rownames(metadata)
+
+# Make sure metadata is factor where needed
+names(metadata) = gsub("Condition..","condition_",names(metadata))
+conditions = names(metadata)[grepl("condition_",names(metadata))]
+for (i in conditions) {
+    metadata[,i] = as.factor(metadata[,i])
+}
+# Load count table for FeatureCounts
```
[the remark this answers is not in the document]
Comment (WackerO) [reply assets/RNAseq_report.Rmd:274]: Ah sorry, on it!

File: assets/RNAseq_report.Rmd [hunk 30]
```diff
@@ -123,21 +248,83 @@ Auf der Morgenstelle 10, 72076 Tuebingen
...
+if (params$input_type == "featurecounts"){
```
[the remark this answers is not in the document]
Comment (WackerO) [reply assets/RNAseq_report.Rmd:275]: I can remove one block I believe, but at least 2 checks are necessary, one for the input stuff and one for the deseq function

File: assets/RNAseq_report.Rmd (hunk 24 above)
Comment (WackerO) [reply assets/RNAseq_report.Rmd:928]: I cannot move the block unless I restructure the actual report. In the block above the code you marked, the heatmaps are put into the report, and above, the PCA is embedded, and both of these blocks require the deseq results

File: assets/RNAseq_report.Rmd (hunk 25 above)
Comment (WackerO) [reply assets/RNAseq_report.Rmd:1187]: Ah yes, I already saw that as well. On it

File: assets/RNAseq_report.Rmd [hunk 31]
```diff
@@ -185,6 +384,302 @@ if ("RIN" %in% colnames(df_QA_neat)){df_QA_neat$RIN <- NULL}
...
+    text(x = c(0, length(h1$counts)), y = 0, label = paste(c(0,1)),
+        adj = c(0.5,1.7), xpd=NA)
+    legend("topleft", fill=rev(colori), legend=rev(names(colori)))
+    dev.off()
+    rm(res,qs,bins,ratios,use,h1,h2,colori)
+}
+```
+<!-- DESeq2 end -->
+
+<!-- metadata end -->
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:677]:
```suggestion
```

File: assets/RNAseq_report.Rmd [hunk 32]
```diff
@@ -407,6 +1208,7 @@ The table for download can be found [here](./differential_gene_expression/final_
 Additional tables for all individual contrasts are provided [here](./differential_gene_expression/DE_genes_tables).
 
 <!-- Show table here removed as I dont like showing only top 20 genes.
+TODO Is this still necessary?
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:1211]: Have you solved this? If the table is not shown, then text line 1202 should be removed

File: assets/RNAseq_report.Rmd [hunk 33]
```diff
@@ -530,27 +1661,29 @@ Hover over the dots to reveal the pathway names.
...
+            next
         }
         q = as.character(DE_genes$Ensembl_ID)
         q_list <- append(q_list, list(q))
         q_names <- append(q_names, fname)
     }
     datasources <- c("KEGG","REAC")
     names(q_list) <- q_names
+    if (!length(q_list)) {
+        stop("list empty?!?")
```
Comment (ggabernet) [inline assets/RNAseq_report.Rmd:1685]: Could be a bit more specific, which list? What does the user need to check?

File: nextflow.config [hunk 34]
```diff
@@ -13,7 +13,7 @@ params {
     metadata =  null
     model = null
     project_summary = null
-    multiqc = null
+    multiqc = 'NO_FILE3'
 
     // Other parameters
     outdir = './results'
```
Comment (ggabernet) [inline nextflow.config:19]: param is missing here

File: nextflow.config [hunk 35]
```diff
@@ -13,7 +13,7 @@ params {
     metadata =  null
     model = null
     project_summary = null
-    multiqc = null
+    multiqc = 'NO_FILE3'
```
Comment (ggabernet) [inline nextflow.config:16]:
```suggestion
    multiqc = 'NO_FILE3'
    versions = null
```

**Review by ggabernet — COMMENTED** [review]
Hi @WackerO , just a handful more comments, then should be good to go!

File: assets/RNAseq_report.Rmd (hunk 33 above)
Comment (WackerO) [reply assets/RNAseq_report.Rmd:1684]: Oops, that is a left-over from some debugging!

**Review by ggabernet — APPROVED** [review]
Looks good to me now :)
